### ISSUE TITLE
GPIO/Button events not working in PiFace implementation/sample

### ISSUE THREAD

**Issue by savageautomate** [Pi4J/pi4j-v1#53]
See comments here:  
http://www.savagehomeautomation.com/projects/raspberry-pi-programming-pi-face-with-java-pi4j.html

See discussion thread here:
http://www.raspberrypi.org/phpBB3/viewtopic.php?f=81&t=35069&p=431572#p431572

And here:
https://groups.google.com/forum/#!topic/pi4j/IlKYpbUXuTI

**Comment by morind79** [comment]
Hi Robert,

Well it helped me, but not completely. Let me explain what I did.
I have a MCP23017 (I2C) on which GPIOB6 and GPIOB7 are set as input.
On these GPIOs, there is buttons.

At the beginning the state of these buttons is unknown, so I read them in my program.
To make test I set them like this B6=false and B7=true.

On my Java program, I just made a loop that read the status of these buttons every 2 seconds.

The result is always the same :

B6=false
B7=false

But if I leave my program runing and change state of one button (B6 for example) then everything is correct.

Really weird.

So using the information you gave me, I added this to my code before the loop :

I2CBus bus = I2CFactory.getinstance(I2CBus.BUS_1);
I2CDevice device = bus.getDevice(0x21);
device.write((byte)0x09, (byte)0xFF);

Then starting with the same state B6=false and B7=true, now on I have : 

1st read (time = 0 second)
B6=false
B7=false
2nd read (time = 2 seconds)
B6=false
B7=true
and continue with the correct read

So it seems the issue is to read the initial state. Have also seen that ?
And I am pretty sure this is the reason why the event I get are "corrupted"...

Thank you

Best regards
Denis

**Comment by savageautomate** [comment]
Hi Denis,

I actually did not make the change to the MCP23S17 GPIO provider class yet.  The PiFace provider uses a different class impl and thats where I made the change.  We should probably look at making the PiFace provider class extend from the MCP23S17 impl to reduce duplication of code.  I would need to compare the two implementations to see exactly what are the differences -- I can't recall now why they are two separate ones to begin with :-)
